#!/usr/bin/python3.7
# -*- coding: utf-8 -*-
# Usage:
# python3.7 ./merge.py file,file1,file2... > merged_file.txt
#
#
import subprocess
from collections import namedtuple
from re import *
from sys import *


def header(date):
    a = 1


def parse(row):
    pattern = compile(r"\s*(?P<rsid>\S+)\s*(?P<chromosome>\S+)\s*(?P<position>\S+)\s*(?P<genotype>\S+)\s*")
    comment = compile(r"\s*[#]\s*.*")
    comment2 = compile(r"\s*RSID\s+CHROMOSOME\s+POSITION\s+RESULT")
    Row = namedtuple('Row', ['rsid', 'chromosome', 'position', 'genotype', 'comment'])
    if comment2.match(row) or comment.match(row):
        return Row(0, 0, 0, 0, comment=True)
    else:
        matches = pattern.search(row)
        return Row(comment=False, rsid=matches.group("rsid"), chromosome=matches.group("chromosome"),
                   position=int(matches.group("position")), genotype=matches.group("genotype"))


if __name__ == '__main__':
    date_str = str(subprocess.check_output(["date"])).strip()
    date_str = date_str.replace("CEST ", "").strip("b").strip("'").strip("n").strip("\\")
    headertxt = "# This data file generated by 23andMe at: " + date_str + "\n#\n# This file contains raw" \
                                                                          "genotype data, including data that is not " \
                                                                          "used in 23andMe reports.\n# This data has " \
                                                                          "undergone a " \
                                                                          "general quality review however only a " \
                                                                          "subset of markers have been \n# " \
                                                                          "individually validated for " \
                                                                          "accuracy. As such, this data is suitable " \
                                                                          "only for research, \n# educational, " \
                                                                          "and informational use " \
                                                                          "and not for medical or other use.\n# \n# " \
                                                                          "Below is a text version of your data.  " \
                                                                          "Fields are " \
                                                                          "TAB-separated\n# Each line corresponds to " \
                                                                          "a single SNP.  For each SNP, we provide " \
                                                                          "its identifier " \
                                                                          "\n# (an rsid or an internal id), " \
                                                                          "its location on the reference human " \
                                                                          "genome, and the \n# genotype " \
                                                                          "call oriented with respect to the plus " \
                                                                          "strand on the human reference sequence.\n# " \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "We are using " \
                                                                          "reference human assembly build 37 (also " \
                                                                          "known as Annotation Release 104).\n# Note " \
                                                                          "that it is " \
                                                                          "possible that data downloaded at different " \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "" \
                                                                          "times may be different due to ongoing \n# " \
                                                                          "improvements " \
                                                                          "in " \
                                                                          "our ability to call genotypes. More " \
                                                                          "information about these changes can be " \
                                                                          "found at:\n# " \
                                                                          "https://you.23andme.com/p/decf709166730166/tools/data/download/\n# \n# More information on " \
                                                                          "reference human assembly build 37 (aka " \
                                                                          "Annotation Release 104):\n# " \
                                                                          "http://www.ncbi.nlm.nih.gov/mapview/map_search.cgi?taxid=9606\n#\n# rsid	chromosome	position	" \
                                                                          "genotype"
    files = argv
    del (files[0])
    if len(files) > 0:
        # first = 0
        values = [set([]) for x in range(25)]  # Sets of positions for the chromosomes {chromosome:(positions)}
        rsids = [{} for x in range(25)]  # {position:rsid}
        genotypes = [{} for x in range(25)]  # {position:genotype}
        from_tr = {'1': 0, '2': 1, '3': 2, '4': 3, '5': 4, '6': 5, '7': 6,
                   '8': 7, '9': 8, '10': 9, '11': 10, '12': 11, '13': 12,
                   '14': 13, '15': 14, '16': 15, '17': 16, '18': 17, '19': 18,
                   '20': 19, '21': 20, '22': 21, 'X': 22, 'Y': 23, 'MT': 24}
        to_tr = {0: '1', 1: '2', 2: '3', 3: '4', 4: '5', 5: '6', 6: '7', 7: '8', 8: '9',
                 9: '10', 10: '11', 11: '12', 12: '13', 13: '14', 14: '15', 15: '16',
                 16: '17', 17: '18', 18: '19', 19: '20', 20: '21', 21: '22',
                 22: 'X', 23: 'Y', 24: 'MT'}
        file_nr = 0
        row_nr = 0
        for file_location in files:
            file_nr += 1
            file_location = file_location.strip()
            with open(file_location, 'r') as input_data_file:
                lines = input_data_file.readlines()
                for line in lines:
                    row_nr += 1
                    the_row = parse(line)
                    if not the_row.comment:
                        chromo = from_tr[the_row.chromosome]
                        pos = the_row.position
                        gtype = the_row.genotype
                        # No reporting conflicts or overlap for now
                        gt = gtype
                        if pos in genotypes[chromo]:
                            current_gt = (genotypes[chromo])[pos]
                            if current_gt == '--':
                                if gtype == '--':
                                    # gt = 'XX'
                                    gt = '--'
                                else:
                                    gt = gtype
                            elif gtype == current_gt or gtype == '--':
                                gt = current_gt
                            elif gtype != current_gt:
                                # print("Conflict at:" + str(pos))
                                gt = 'XX'
                        else:
                            values[chromo].add(pos)
                        (genotypes[chromo])[pos] = gt
                        (rsids[chromo])[pos] = the_row.rsid
    print(headertxt)
    for i in range(25):
        lst = sorted(values[i])
        for pos in lst:
            # rsid	chromosome	position	genotype
            gt = (genotypes[i])[pos]
            if gt != 'XX' and gt != '--':
                outp = str((rsids[i])[pos]) + "\t" + to_tr[i] + "\t" + str(pos) + "\t" + str(gt) + "\t"
                print(outp)

    exit(0)
else:
    print('This requires some input files.')
    exit(1)


# except:
# print(repr(the_row))
# print(repr(line))
# print("row: " + str(row_nr))
# print("file: " + str(file_nr))
# exit(1)
